lib/web_of_science/contributions.rb

```diff
@@ -50,14 +49,28 @@ def find_or_create_contribution(author, publication)
 
     # Does record have a contribution for this author? (based on matching PublicationIdentifiers)
     # Note: must use unique identifiers, don't use ISSN or similar series level identifiers
+    # We search for all PubIDs at once instead of serial queries.  No need to hit the same table multiple times.
     # @param [Author] author
     # @param [WebOfScience::Record] record
-    # @return [Boolean]
+    # @return [::Contribution, nil] a matched or newly minted Contribution
     def found_contribution?(author, record)
-      contribution_by_identifier?(author, 'WosUID', record.uid) ||
-        contribution_by_identifier?(author, 'WosItemID', record.wos_item_id) ||
-        contribution_by_identifier?(author, 'doi', record.doi) ||
-        contribution_by_identifier?(author, 'pmid', record.pmid)
+      pub = Publication.joins(:publication_identifiers).where(
+        "publication_identifiers.identifier_value IS NOT NULL AND (
+         (publication_identifiers.identifier_type = 'WosUID' AND publication_identifiers.identifier_value = ?) OR
+         (publication_identifiers.identifier_type = 'WosItemID' AND publication_identifiers.identifier_value = ?) OR
+         (publication_identifiers.identifier_type = 'doi' AND publication_identifiers.identifier_value = ?) OR
+         (publication_identifiers.identifier_type = 'pmid' AND publication_identifiers.identifier_value = ?))",
+         record.uid, record.wos_item_id, record.doi, record.pmid
+      ).order(
+        "CASE
+          WHEN publication_identifiers.identifier_type = 'WosUID' THEN 0
+          WHEN publication_identifiers.identifier_type = 'WosItemID' THEN 1
+          WHEN publication_identifiers.identifier_type = 'doi' THEN 2
+          WHEN publication_identifiers.identifier_type = 'pmid' THEN 3
+         END"
+      ).first
+      return unless pub
+      find_or_create_contribution(author, pub)
     end
```

Comment:
Does this version of the query work when any of `record.wos_item_id, record.doi, record.pmid` are `nil`?  AFAIK, the `record.uid` should be guaranteed by the WOS-API.

Reply:
Any `nil` value won't matter because they will be excluded by the first clause: `publication_identifiers.identifier_value IS NOT NULL`.  Can't match a `NOT NULL` condition and a `= NULL` condition at the same time.

`.first` does `LIMIT 1`, as you can verify in your local rails dev console.  

I'm assuming we cared based on the priority of the previous implementation.  I reproduced that priority for the unified query.

Reply:
The priority of the previous query was on the match-priority, which resulted in a return priority.  AFAICT, the new query doesn't impose a match-priority up front, but sorts the results by that priority.  It gets all the results in one hit, to get one result.  So the change in the query does hit the db once and that seems like a performance improvement in terms of IO perhaps - but the db search algorithm itself might have potentially more work to do on the query, depending on the case?  If it matched on say WOS-UID, the previous code was faster.  If it matched on the last identifier, the new code could be faster.  The table is indexed by type & value, so that helps.  I'm not sure what the details of the execution times could be - it would be interesting to benchmark it.  On this branch, did the profiling results improve for this change?

Reply:
I would prefer the combined query approach since the columns we are querying on are indexed - mysql should be well optimized to do this plus we avoid more ActiveRecord overhead.

Reply:
The query may pull more rows prior to the SORT and LIMIT.  But we if we aren't worried about the performance of the DB pulling a single row based on two indexed fields, we shouldn't suddenly become worried when that is increased to MAX 4 rows.